### ISSUE TITLE
Scripts/hooks 

### ISSUE THREAD

**Issue by mthenw** [serverless/serverless#3010]
# This is a Feature Proposal

## Description

* What is the use case that should be solved. The more detail you describe this in the easier it is to understand for us.

Extensibility. It's pretty common use case to run some build/compile tools before deploy. Babel/Webpack/Java/.Net tools. The only way to that right now is to use plugins or run them manually. My proposition is to enable running scripts on hooks (not those exposed by plugins, more general like pre-package, pre-deploy, post-deploy). It's similar to npm scripts. 

* If there is additional config how would it look

```
service: test

provider:
...

scripts:
  pre-package: babel script.js --out-file script-compiled.js
  post-deploy: curl http://deploy.done/

functions:
...
```

**Comment by pmuens** [comment]
Neat idea! I like it!

Should we incorporate the lifecycle events here? Like

```yaml
before:deploy: node foo.js
after:deploy: echo "success"
```

This way developers don't have to learn new concepts / syntax and it could be fairly easy to add this functionality as it maps to the lifecycles we already have.

**Comment by DavidWells** [comment]
I like the [npm pre/post conventions](https://docs.npmjs.com/misc/scripts) "Pre and post commands with matching names will be run for those as well (e.g. premyscript, myscript, postmyscript)"

```
{
 prefoo: "echo before foo",
 foo: "echo 'foo'",
 postfoo: "echo do thing after foo"
}
```

But either way, this feature would be pretty nice to have =)

**Comment by mthenw** [comment]
> This way developers don't have to learn new concepts / syntax and it could be fairly easy to add this functionality as it maps to the lifecycles we already have.

yeah... Not sure about exposing internal plugins hooks outside. Also they are pretty granular, I don't think it's needed. Definitely something to discuss.

**Comment by mvila** [comment]
I also think that having a simple way to define commands and hooks would be a nice addition.

I propose the following configuration:

```yaml
scripts:
  commands:
    hello: echo Hello from ${self:service} service!
  hooks:
    'deploy:createDeploymentArtifacts': npm run compile
```

I have implemented this feature in the following plugin:

https://github.com/mvila/serverless-plugin-scripts

**Comment by pmuens** [comment]
Closing since there's a plugin which provides a solution for that.

**Comment by AntonOellerer** [comment]
I am afraid the plugin has gone into [maintainance mode](https://github.com/mvila/serverless-plugin-scripts/issues/10).
Is there any way either the feature proposal or the plugin could be brought back to life?
